catkit2/services/thorlabs_cube_motor/thorlabs_cube_motor.py

```diff
@@ -0,0 +1,113 @@
+from catkit2.testbed.service import Service
+
+import thorlabs_apt as apt
+import numpy as np
+
+
+class ThorlabsCubeMotor(Service):
+    """
+    Service for controlling a Thorlabs cube motor.
+
+    This service has been tested with the TDC001 and KDC101 cube motor controllers.
+    """
+
+    def __init__(self):
+        super().__init__('thorlabs_cube_motor')
+
+        self.cube_serial_number = self.config['serial_number']
+
+        self.command = self.make_data_stream('command', 'float64', [1], 20)
+        self.current_position = self.make_data_stream('current_position', 'float64', [1], 20)
+
+        self.motor = None
+        self.cube_model = None
+        self.motor_unit = None
+        self.motor_max = None
+        self.motor_min = None
+
+    def open(self):
+        """
+        Open connection to the motor.
+
+        This will also check if the configured characteristics are matching the one the cube have.
+        """
+        self.motor = apt.Motor(self.cube_serial_number)
+
+        # Retrieve the min, max, unit and model from the connected device.
+        min = self.motor.get_stage_axis_info()[0]
+        max = self.motor.get_stage_axis_info()[1]
+        unit = None
+        model = self.motor.hardware_info[0][2:-1]
+        if self.motor.get_stage_axis_info()[2] == 1:
+            unit = 'mm'
+        if self.motor.get_stage_axis_info()[2] == 2:
+            unit = 'deg'
+
+        # Read min, max, unit and model from service configuration.
+        self.min_pos = self.config['motor_min_pos']
+        self.max_pos = self.config['motor_max_pos']
+        self.unit = self.config['motor_unit']
+        self.cube_model = self.config['cube_model']
+
+        # Compare the device parameters to the service configuration.
+        if not (self.min_pos == min and self.max_pos == max and self.unit == unit and self.cube_model == model):
+            raise ValueError("Device parameters don't match configuration parameters.")
+
+    def main(self):
+        while not self.should_shut_down:
+            try:
+                # Get an update for the motor position.
+                frame = self.command.get_next_frame(10)
+                # Set the current position if a new command has arrived.
+                self.set_current_position(frame.data[0])
+            except Exception:
+                # Timed out. This is used to periodically check the shutdown flag.
+                continue
+
+    def close(self):
+        apt._cleanup()

```

Comment:
I am unsure of this behavior when we have several motor services running. Also, do we meet to call the init on the apt library? Might be a follow-up to keep in mind, and keep this PR just to the development of a single service.

See also https://github.com/qpit/thorlabs_apt/issues/6

Reply:
This differs from implementation to implementation. Some devices require having only a single device loaded at the same time, others allow for multiple. If the documentation doesn't tell you, trying it out is the only way to know.